src/server/worker/logs/logger.go

```diff
@@ -0,0 +1,277 @@
+func NewStatlessLogger(pipelineInfo *pps.PipelineInfo) TaggedLogger {
+	return newLogger(pipelineInfo)
+}
+
+// NewMasterLogger constructs a TaggedLogger for the given pipeline that
+// includes the 'Master' flag.  This is typically used for all logging from the
+// worker master goroutine.
+func NewMasterLogger(pipelineInfo *pps.PipelineInfo) TaggedLogger {
+	result := newLogger(pipelineInfo) // master loggers don't log stats
+	result.template.Master = true
+	return result
+}
+
+// WithJob clones the current logger and returns a new one that will include
+// the given job ID in log statement metadata.
+func (logger *taggedLogger) WithJob(jobID string) TaggedLogger {
+	result := logger.clone()
+	result.template.JobID = jobID
+	return result
+}
+
+// WithJob clones the current logger and returns a new one that will include
+// the given data inputs in log statement metadata.
+func (logger *taggedLogger) WithData(data []*common.Input) TaggedLogger {
+	result := logger.clone()
+
+	// Add inputs' details to log metadata, so we can find these logs later
+	result.template.Data = []*pps.InputFile{}
+	for _, d := range data {
+		result.template.Data = append(result.template.Data, &pps.InputFile{
+			Path: d.FileInfo.File.Path,
+			Hash: d.FileInfo.Hash,
+		})
+	}
+
+	// This is the same ID used in the stats tree for the datum
+	result.template.DatumID = common.DatumID(data)
+	return result
+}
+
+// WithUserCode clones the current logger and returns a new one that will
+// include the 'User' flag in log statement metadata.
+func (logger *taggedLogger) WithUserCode() TaggedLogger {
+	result := logger.clone()
+	result.template.User = true
+	return result
+}
+
+// JobID returns the current job that the logger is configured with.
+func (logger *taggedLogger) JobID() string {
+	return logger.template.JobID
+}
+
+func (logger *taggedLogger) clone() *taggedLogger {
+	return &taggedLogger{
+		template:     logger.template,  // Copy struct
+		stderrLog:    logger.stderrLog, // logger should be goroutine-safe
+		marshaler:    &jsonpb.Marshaler{},
+		putObjClient: logger.putObjClient,
+		msgCh:        logger.msgCh,
+	}
+}
+
+// Logf logs the line Sprintf(formatString, args...), but formatted as a json
+// message and annotated with all of the metadata stored in 'loginfo'.
+//
+// Note: this is not thread-safe, as it modifies fields of 'logger.template'
+func (logger *taggedLogger) Logf(formatString string, args ...interface{}) {
+	logger.template.Message = fmt.Sprintf(formatString, args...)
+	if ts, err := types.TimestampProto(time.Now()); err == nil {
+		logger.template.Ts = ts
+	} else {
+		logger.Errf("could not generate logging timestamp: %s\n", err)
+		return
+	}
+	msg, err := logger.marshaler.MarshalToString(&logger.template)
+	if err != nil {
+		logger.Errf("could not marshal %v for logging: %s\n", &logger.template, err)
+		return
+	}
+	fmt.Println(msg)
+	if logger.putObjClient != nil {
+		logger.msgCh <- msg + "\n"
+	}
+}
```

Comment:
I'm not quite sure where if it goes here or in `Errf` but I think it makes sense to extract stacks from errors in these functions and embed them in the logged messages.